bundles/org.openhab.binding.vizio/src/main/java/org/openhab/binding/vizio/internal/handler/VizioHandler.java

```diff
@@ -0,0 +1,572 @@
+import org.openhab.core.library.types.StringType;
+import org.openhab.core.thing.Channel;
+import org.openhab.core.thing.ChannelUID;
+import org.openhab.core.thing.Thing;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.ThingStatusDetail;
+import org.openhab.core.thing.binding.BaseThingHandler;
+import org.openhab.core.types.Command;
+import org.openhab.core.types.RefreshType;
+import org.openhab.core.types.StateOption;
+import org.openhab.core.types.UnDefType;
+import org.osgi.framework.FrameworkUtil;
+import org.osgi.framework.ServiceRegistration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonSyntaxException;
+
+/**
+ * The {@link VizioHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Michael Lobstein - Initial contribution
+ */
+@NonNullByDefault
+public class VizioHandler extends BaseThingHandler {
+    private final Logger logger = LoggerFactory.getLogger(VizioHandler.class);
+    private final HttpClient httpClient;
+    private final VizioStateDescriptionOptionProvider stateDescriptionProvider;
+    private final String dbAppsJson;
+
+    private @Nullable ServiceRegistration<?> serviceRegistration;
+    private @Nullable ScheduledFuture<?> refreshJob;
+    private @Nullable ScheduledFuture<?> metadataRefreshJob;
+
+    private VizioCommunicator communicator;
+    private List<VizioApp> userConfigApps = new ArrayList<VizioApp>();
+    private Object sequenceLock = new Object();
+
+    private int pairingDeviceId = -1;
+    private int pairingToken = -1;
+    private Long currentInputHash = 0L;
+    private Long currentVolumeHash = 0L;
+    private String currentApp = EMPTY;
+    private String currentInput = EMPTY;
+    private boolean currentMute = false;
+    private int currentVolume = -1;
+    private boolean powerOn = false;
+    private boolean debounce = true;
+
+    public VizioHandler(Thing thing, HttpClient httpClient,
+            VizioStateDescriptionOptionProvider stateDescriptionProvider, String vizioAppsJson) {
+        super(thing);
+        this.httpClient = httpClient;
+        this.stateDescriptionProvider = stateDescriptionProvider;
+        this.dbAppsJson = vizioAppsJson;
+        this.communicator = new VizioCommunicator(httpClient, EMPTY, -1, EMPTY);
+    }
+
+    @Override
+    public void initialize() {
+        logger.debug("Initializing Vizio handler");
+        final Gson gson = new Gson();
+        VizioConfiguration config = getConfigAs(VizioConfiguration.class);
+
+        @Nullable
+        String host = config.hostName;
+        final @Nullable String authToken = config.authToken;
+        @Nullable
+        String appListJson = config.appListJson;
+
+        if (host == null || host.isEmpty()) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Host Name must be specified");
```

Comment:
Please add an i18n string for this message. See https://www.openhab.org/docs/developer/utils/i18n.html#using-custom-keys